### PR TITLE
[SYSTEMML-2004] Covariance Kernels

### PR DESCRIPTION
**Pull request by j143**
_what is here?_
**covariance kernel**
1. a covariance matrix converts input values. 
2. in one dimension fn changes in 100 units, while in the other fn changes in 1 unit of the scale.

### CODE REVIEW

**Comment by j143** [comment]
Hi @niketanpansare , this covariance is one of the key parts of the implementation. This preliminary one have some loops. 

This is one of the places, I would like to take the help. Thanks. 😃

**Comment by akchinSTC** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2243/

**Comment by niketanpansare** [comment]
I would recommend that you contact Prithvi or Sasha as they have worked on http://apache.github.io/systemml/algorithms-descriptive-statistics.html scripts and use  https://github.com/apache/systemml/blob/master/scripts/algorithms/bivar-stats.dml and https://github.com/apache/systemml/blob/master/scripts/algorithms/Univar-Stats.dml as reference :)

**Comment by j143** [comment]
**1. let us consider `r^2(x, x')` in the below formula:**
 for given dimension `d`, the the combination(substract!) of each element with every other element.

**2. for example:** 
`d = 2;`
`X = matrix(rows=n, cols=d);`

1. `X(1,d) - X(1,d)` will be the `(1,1)`th element
2. `X(1,d) - X(2,d)` will be the `(1,2)`th element
3. `X(2,d) - X(1,d)` will be the `(2,1)`th element and so on..

[ref: line no. 35 to 55](https://github.com/apache/systemml/pull/719/files#diff-f24dad2d9811c6f4cb73977436e717b5R36)

**3. Formula**:
se_kernel

File: scripts/staging/gaussian_process/covariance.dml [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+# specific language governing permissions and limitations
+# under the License.
+#
+#-------------------------------------------------------------
+
+
+cov = function(matrix[double] X)
+  return(matrix[double] out) {
+
+  K = -2 * X %*% t(X) + rowSums(X^2) + t( rowSums(X^2) )
```
Comment (prithvirajsen) [inline scripts/staging/gaussian_process/covariance.dml]: Unless SystemML takes care of the CSE (`rowSums(X^2)`), consider temp-ing it to a variable and reusing.

**Comment by prithvirajsen** [comment]
LGTM

**Comment by j143** [comment]
Thanks @prithvirajsen , we will try to optimize this preliminary script, once the whole feature gets into a shape. 

Hi @niketanpansare , test is done here with calculation of cholesky, which verifies the properties of `cov()`. If the things are OK, can this be merged. Thanks.

**Comment by niketanpansare** [comment]
Thanks @prithvirajsen and @j143 ... I will merge this PR today :+1:

**Pull request closed without merging**